File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];

        for (int i = 0; i < n; i++) {
            ans[i] = 1;
        }
        int left = 1;
        for (int i =0; i<n; i++){
            ans[i]=left;
            left *= nums[i];
        }
        int right=1;
        for (int j= n-1; j>=0; j--){
            ans[j] *= right;
            right *= nums[j];
        }
        return ans;

    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        length = len(nums)
        answer =[1]*length
        for i in range(1,length):
            answer[i] = nums[i-1]*answer[i-1]
        
        R =1
        for i in reversed(range(length)):
            answer[i] = answer[i]*R
            R*=nums[i]
        return answer        
        